### DISCUSSION TITLE
Support templating of the searchbox web part

### DISCUSSION THREAD

**Discussion by edarroudi** [microsoft-search/pnp-modern-search#1470] [Q&A] [answered]
I would like to change some css and add additional controls to the searchbox webpart. Can we have here a custom template as in the search results / filter webpart? 
I'll implement this myself following the structure of the other templating webparts.
Of course i won't do it if you won't integrate it :)

**Comment by wobba** [comment, marked as answer]
I'm not opposed to it. Can you share what kind of settings it would have, and how do you want to handle the suggestion/not-suggestion toggle which today are two controls.

**Reply by edarroudi** [reply]
@wobba Ehm did you comment on the wrong discussion? Did you mean that? https://github.com/microsoft-search/pnp-modern-search/discussions/1476

**Reply by edarroudi** [reply]
Oh or are you talking about the Suggestion Provider?

**Reply by wobba** [reply]
No..I'm talking about having a template for search box. See \search-parts\src\webparts\searchBox\components\SearchBoxContainer.tsx L27 and L39.

**Reply by edarroudi** [reply]
Personally i don't see a big issue there with the toggle as long as we export both as web-components which the user can use.
the template would then contain something like this:
 
` {{#if @root.properties.enableQuerySuggestions}}
<pnp-autocomplete-searchbox>
{{else}}
<pnp-searchbox>
{{/if}}`

of course we have to autobind somehow the events, maybe restrict them so the users can't set them?

I will give it some additonal thought and come up with a plan and present it :) If you have additional thoughts or requirements i would like to hear them.

**Reply by wobba** [reply]
I don't see an issue either as long as it's all done correct. As long as we document how the toggle can impact a custom template it should work.

**Reply by FranckyC** [reply]
Hi there, sorry to intervene but I'm not in favor of this feature as it is already possible to achieve the same goal using a custom web component directly in the search results and use the '#' URL parameter to pass the keyword value to other components. 

Adding template support to the search box introduces unnecessary complexity to the solution and is not an essential feature in my opinion.

**Reply by edarroudi** [reply]
Hi @FranckyC , 

You are talking about just using a result webpart, creating a textbox component that changes the url (# or query) and add it to the resultwebpart. So the resultwebpart would work as searchbox and resultwebpart. 

I would need to copy the features of the searchBox Webpart (like SuggestionProviders) to the new webcomponent. And loose the property interface.

I probably can't argue against this. Even if i show you a prototype of the implementation, i won't be able to change your mind?

**Reply by FranckyC** [reply]
> I would need to copy the features of the searchBox Webpart (like SuggestionProviders) to the new webcomponent. And loose the property interface.

> I would like to change some css and add additional controls to the searchbox webpart.

Can you explain the requirement a little bit more? Web components are just wrappers so you can put pretty much anything here. Except the layout flexibility, I don't really understand why you want to update the exiting search box.

**Reply by edarroudi** [reply]
I understand what webcomponents are, but as i still would have to modify the existing <SearchBox> / <SearchBoxAutoComplete> controls to allow some sort of CSS modification, i would have to modify that code in my custom controls as i'm not able to "officially" modify those controls.

What i want to achieve is something like this:
![image](https://user-images.githubusercontent.com/935493/140754227-7f4edcae-7c95-459f-b45b-3c3447a869fe.png)
I understand you might want to do something like this through Verticals but my customer has different requests.

Or maybe i want to additionally have a dropdown there for some purpose. Of course not everything makes sense but i think some people want to have some kind of control of how the textbox is rendered and mabe the possibility to add additional controls ...

**Reply by FranckyC** [reply]
The requirement you describe seems very specific to your project. I'm not convinced this should be implemented in the public solution. The risk of breaking changes/regressions/maintenance burdnown is too high compared to the provided value. Also, this feature has never been discussed in almost 4 years of moderns search solution so I guess this is not a top priority feature for people. Finally, as I said earlier, there are workarounds to adress this specific scenario with existing components.

Hope you understand my position.

**Reply by wobba** [reply]
Another solution is to create your own search box with needed functionality, and make it work with the PnP parts via dynamic data.

**Reply by FranckyC** [reply]
Good alternative @wobba, you could use the extensibility repo for this purpose https://github.com/microsoft-search/pnp-modern-search-extensibility-samples

**Reply by edarroudi** [reply]
Hey guys, thanks for the feedback. I totally understand your reasoning - even if i have another view on this topic. But i'll try to use your suggestions and also continue to think of another minimum "layouting" solution on my end. Thanks again! This can be closed.
